hngprojects/hng_boilerplate_nestjs
# feat: create and fetch languages

## What does the PR do?
This PR is about creating two endpoints for adding new languages and returning a list of supported system languages
## How should this be manually tested?
**Send a post request to `/api/v1/language` with a request body as below**
```json
{
    "code": string,
    "name": "string,
    "native_name": string
    "direction": "LTR | RTL"
  }

```

- Successful Response

On a successful  language creation, the API should return a 201  status code.
The response body should contain the newly created language  details.
```
{
  "status_code": 201,
  "message": "Language Created Successfully",
  "language": {
    "code": "ar1236789915",
    "name": "Arabic1",
    "native_name": "العربية",
    "direction": "LTR",
    "id": "bd3605ba-ef03-4d03-931a-4b97c82f565e",
    "created_at": "2024-07-24T15:33:05.460Z",
    "updated_at": "2024-07-24T15:33:05.460Z"
  }
}
```

- Authentication Error

When the current user is not properly authenticated to fetch or create a new language
```
{
  "message": "User is currently unauthorized, kindly authenticate to continue",
  "status_code": 401
}
```

**Send a get request to `/api/v1/language` with an authentication token**
- Successful Response

The API should return a 200 status code on a successful language fetch.
The response body should contain the fetched language details.
```
{
  "status_code": 200,
  "message": "Success",
  "languages": [
    {
      "id": "345c4d40-78dd-4269-a05e-f8aa7a2f6d25",
      "code": "ar",
      "name": "Arabic",
      "native_name": "العربية",
      "direction": "RTL",
      "created_at": "2024-07-24T13:09:58.439Z",
      "updated_at": "2024-07-24T13:09:58.439Z"
    },
    {
      "id": "b6bd5cb3-140b-4a46-bb06-6192aba0b2a8",
      "code": "ar123",
      "name": "Arabic1",
      "native_name": "العربية",
      "direction": "RTL",
      "created_at": "2024-07-24T13:10:53.907Z",
      "updated_at": "2024-07-24T13:10:53.907Z"
    },
    {
      "id": "f06d6ca0-0496-40da-aa4d-fd681e6b3207",
      "code": "ar1236",
      "name": "Arabic1",
      "native_name": "العربية",
      "direction": "RTL",
      "created_at": "2024-07-24T13:11:03.244Z",
      "updated_at": "2024-07-24T13:11:03.244Z"
    },
    
  ]
}
```


## Checklist of what you did
- [x] Implemented endpoint to create a language
- [x] Validated request body and returned validation errors when validation fails
- [x] Prevented duplicate record by making language code unique
- [x] Created a language once user is authenticated
- [x] Created a record for language in the database
- [x] Returned success message with response payload
- [x] Added unit tests for language service
## Reference Issue
[FEAT]: Support for Language Selection Feature #34](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/34)
